### PR TITLE
Extrae initialization and runtime hook for Distributed workflow

### PR DESCRIPTION
**Pull request by clasqui**
This PR adds the functionality to start workers that overdub the runtime's main event loop with a custom `addprocs_extrae` function.

### CODE REVIEW

File: test/test-distributed-work.jl [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    Cassette.overdub(Extrae.ExtraeCtx(), remote_do, Cassette.overdub, 2, Extrae.ExtraeCtx(), random_sleep)
+    # Cassette.overdub(Extrae.ExtraeCtx(), remote_do, Cassette.overdub, 3, Extrae.ExtraeCtx(), random_sleep)
+    # Cassette.overdub(Extrae.ExtraeCtx(), remote_do, Cassette.overdub, 4, Extrae.ExtraeCtx(), random_sleep)
+    sleep(10)
+    @everywhere Extrae.finish()
+end
+
+test_distributed_work()
+
+println("END TEST")
```
Comment (mofeing) [inline test/test-distributed-work.jl:27]: This is not how Julia tests are supposed to work 😅 
But I think we can do something.

File: src/Instrumentation/ExtraeLocalManager.jl [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+function Distributed.manage(manager::ExtraeLocalManager, id::Integer, config::WorkerConfig, op::Symbol)
+    if op === :interrupt
+        kill(config.process, 2)
+    end
+end
+
+export launch
+
+export manage
+
```
Comment (mofeing) [inline src/Instrumentation/ExtraeLocalManager.jl:54]:
```suggestion
```
You don't have to export them. They are already exported by `Distributed`.

File: src/API.jl [hunk 3]
```diff
@@ -27,7 +27,23 @@ This routine is called automatically in different circumstances, which include:
...
+  ## TODO: This setup should depend on isntrumentation options.
+  ## For example, if isntrumenting Distributed, here we setup the
+  ## Distributed functions to identify resources
+  FFI.Extrae_set_numtasks_function(dist_numtasks)
+  FFI.Extrae_set_taskid_function(dist_taskid)
+
+  ## Setup traceid for not intereference
+  name = "JULIATRACE" * string(Distributed.myid())
+  var = "EXTRAE_PROGRAM_NAME"
+  @ccall setenv(var::Cstring, name::Cstring)::Cint
```
Comment (mofeing) [inline src/API.jl:41]:
```suggestion
  ENV["EXTRAE_PROGRAM_NAME"] = "JULIATRACE$(Distributed.myid())"
```

File: src/Instrumentation/Distributed.jl [hunk 4]
```diff
@@ -53,3 +56,24 @@ Cassette.posthook(::ExtraeCtx, _, ::typeof(process_messages), args...) = emit(Di
...
+    id = Distributed.myid() - 1
+    return id
+end
+export dist_taskid
+
+function dist_numtasks()::Cuint
+    nworkers = Distributed.nworkers() + 1
+    return nworkers
+end
+export dist_numtasks
```
Comment (mofeing) [inline src/Instrumentation/Distributed.jl:71]:
```suggestion
dist_taskid()::Cuint = Distributed.myid() - 1

dist_numtasks()::Cuint = Distributed.nworkers() + 1
```

File: src/Instrumentation/ExtraeLocalManager.jl [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+function Distributed.launch(manager::ExtraeLocalManager, params::Dict, launched::Array, c::Condition)
+    dir = params[:dir]
+    exename = params[:exename]
+    exeflags = params[:exeflags]
+    bind_to = manager.restrict ? `127.0.0.1` : `$(LPROC.bind_addr)`
+    cookie = cluster_cookie()
+
+    #hookline = """using Distributed; f = Base.open("hooked.txt", "w"); write(f, $(repr(string(cookie))))"""
+    hookline = """using Distributed; using Extrae; using Cassette; Cassette.overdub(Extrae.ExtraeCtx(), start_worker, $(repr(string(cookie))))"""
```
Comment (mofeing) [inline src/Instrumentation/ExtraeLocalManager.jl:23]:
```suggestion
    hookline = """
        using Distributed
        using Extrae
        using Cassette
        
        Cassette.overdub(Extrae.ExtraeCtx(), start_worker, $(repr(string(cookie))))
        """
```

File: test/test-distributed-work.jl (hunk 1 above)
Comment (clasqui) [reply test/test-distributed-work.jl:27]: I know, it is just a fast way for me to try things. We can put it somewhere else!!

Comment (mofeing) [reply test/test-distributed-work.jl:27]: In that case, I like putting this kind of codes inside the `scripts/` folder. Do you mind moving it there?

File: src/Instrumentation/ExtraeLocalManager.jl [hunk 6]
```diff
@@ -0,0 +1,57 @@
+"""
+ExtraeLocalManager.jl
+Implements a copy of the default LocalClusterManager that starts
+workers with an overdubbed event loop.
+
+"""
+
```
Comment (mofeing) [inline src/Instrumentation/ExtraeLocalManager.jl:7]:
```suggestion
"""
    ExtraeLocalManager

Implements a copy of the default `LocalClusterManager` that starts
workers with an overdubbed event loop.
"""

```

File: src/Instrumentation/ExtraeLocalManager.jl [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+struct ExtraeLocalManager <: ClusterManager
+    np::Int
+    restrict::Bool  # Restrict binding to 127.0.0.1 only
+end
+
+Base.show(io::IO, manager::ExtraeLocalManager) = print(io, "ExtraeLocalManager()")
```
Comment (mofeing) [inline src/Instrumentation/ExtraeLocalManager.jl:13]:
```suggestion
Base.show(io::IO, manager::ExtraeLocalManager) = print(io, "ExtraeLocalManager(#procs=$(manager.np), restrict=$(manager.restrict))")
```

File: src/Instrumentation/ExtraeLocalManager.jl [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+function Distributed.launch(manager::ExtraeLocalManager, params::Dict, launched::Array, c::Condition)
+    dir = params[:dir]
+    exename = params[:exename]
+    exeflags = params[:exeflags]
+    bind_to = manager.restrict ? `127.0.0.1` : `$(LPROC.bind_addr)`
+    cookie = cluster_cookie()
+
+    #hookline = """using Distributed; f = Base.open("hooked.txt", "w"); write(f, $(repr(string(cookie))))"""
```
Comment (mofeing) [inline src/Instrumentation/ExtraeLocalManager.jl:22]:
```suggestion
```

File: src/Instrumentation/ExtraeLocalManager.jl [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    hookline = """
+        using Distributed
+        using Extrae
+        using Cassette
+        
+        Cassette.overdub(Extrae.ExtraeCtx(), start_worker, $(repr(string(cookie))))
```
Comment (mofeing) [inline src/Instrumentation/ExtraeLocalManager.jl:28]: Are both `repr` and `string` needed? Either use or another no?

File: src/Instrumentation/Distributed.jl [hunk 10]
```diff
@@ -53,3 +89,39 @@ Cassette.posthook(::ExtraeCtx, _, ::typeof(process_messages), args...) = emit(Di
 
 Cassette.prehook(::ExtraeCtx, ::typeof(interrupt), args...) = emit(DistributedInterrupt)
 Cassette.posthook(::ExtraeCtx, _, ::typeof(interrupt), args...) = emit(DistributedEnd)
+
+# workload execution
+Cassette.prehook(::ExtraeCtx, ::typeof(Distributed.run_work_thunk), f::Function, args...) = emit(DistributedUsefulWork)
+Cassette.posthook(::ExtraeCtx, _, ::typeof(Distributed.run_work_thunk), f::Function, args...) = emit(DistributedNotUsefulWork)
```
Comment (mofeing) [inline src/Instrumentation/Distributed.jl:95]: If you are not dispatching on an argument, not typing is best practice.

```suggestion
Cassette.prehook(::ExtraeCtx, ::typeof(Distributed.run_work_thunk), f, args...) = emit(DistributedUsefulWork)
Cassette.posthook(::ExtraeCtx, _, ::typeof(Distributed.run_work_thunk), f, args...) = emit(DistributedNotUsefulWork)
```

**Review by mofeing — CHANGES_REQUESTED** [review]
Just requesting some small changes and it should be done.

File: src/API.jl [hunk 11]
```diff
@@ -77,6 +100,7 @@ description(::E) where {E<:Event} = description(E)
 Add a single timestampted event into the tracefile.
 """
 function emit(::Event{T,V}; counters::Bool=false) where {T,V}
+    println("Event emit: $(T): $(V)")
```
Comment (mofeing) [inline src/API.jl:103]: I don't think this is a good idea... Maybe you should replace it with `Logging.@debug`

File: src/Instrumentation/ExtraeLocalManager.jl (hunk 9 above)
Comment (clasqui) [reply src/Instrumentation/ExtraeLocalManager.jl:28]: I see

File: src/Instrumentation/Distributed.jl (hunk 10 above)
Comment (clasqui) [reply src/Instrumentation/Distributed.jl:95]: I am dispatching on an argument. The run work thunk function has two methods and right now, for this event I want to only capture the one that has Function as first argument  type, not the other.

File: src/API.jl (hunk 11 above)
Comment (clasqui) [reply src/API.jl:103]: Yep, looks like a useful function.

**Pull request merged**